### DISCUSSION TITLE
Authentication AppOnly on Linux

### DISCUSSION THREAD

**Discussion by ChrTall** [pnp/pnpcore#613] [Q&A] [answered]
I can not find a way to authenticate with Client Id, Tenant Id and Client Secret.
Only with a Certificate, which is also a bit confusing as you have to enter StoreName and StoreLocation which seems to be Windows only if I am not mistaken. I have generated a certificate with openssl and uploaded it to AAD. Can you explain how I can do this on Linux? Do I have to import the Certificate in my native keychain?

**Comment by jansenbe** [comment, marked as answer]
Hi @ChrTall ,

The reason why you need a certificate is because it's required when using app-only for "classic" SharePoint APIs like SPO REST and CSOM. Since PnP Core SDK always uses SPO REST it implies you need a certificate for getting app-only to work.

.NET Core and .NET 5 on Linux do support certificate handling, the available stores are slightly different though. See https://docs.microsoft.com/en-us/dotnet/standard/security/cross-platform-cryptography#x509store for information about cert stores in .NET. Also https://newbedev.com/where-does-net-core-search-for-certificates-on-linux-platform is a good resource.

Next to adding the cert into the store you can you can also in your code read the cert from your PFX file and then pass that in. See https://pnp.github.io/pnpcore/api/PnP.Core.Auth.X509CertificateAuthenticationProvider.html#PnP_Core_Auth_X509CertificateAuthenticationProvider__ctor_System_String_System_String_X509Certificate2_ for the needed `X509CertificateAuthenticationProvider` constructor. See https://pnp.github.io/pnpcore/using-the-sdk/configuring%20authentication.html#configuring-pnp-core-sdk-to-use-the-configured-application-1 for a sample showing how to use `X509CertificateAuthenticationProvider` via code.

**Reply by ChrTall** [reply]
Thank you for the detailed information/help. I was able to import the pfx file into the store with this dotnet tool:
https://github.com/gsoft-inc/dotnet-certificate-tool
For others with the same problem:
It is not possible to simply copy the pfx file into ~/.dotnet/corefx/cryptography/x509stores/, you have to import it with the tool.
Also installing powershell on Linux, in order to run the powershell script does not work because it is missing the PKI Module which is only available on Windows.
In the future, it would be helpful to provide a bash script in the Documentation, which is similiar to the powershell script.

**Reply by jansenbe** [reply]
Thanks for sharing your learnings @ChrTall. I'll see how these can be integrated in our docs to make this easier for future users.
